src/lib/api/projects.ts

```diff
@@ -104,40 +110,223 @@ export async function getShared(
  * it returns the updated project object.
  */
 export async function pinProject(
-  csrftoken: string,
   id: number,
   pinned = true,
+  csrf_token: string,
   fetch = globalThis.fetch,
 ): Promise<Project> {
   const endpoint = new URL(`projects/${id}/`, BASE_API_URL);
   const options: RequestInit = {
     credentials: "include",
     method: "PATCH", // this component can only update whether a project is pinned
     headers: {
-      [CSRF_HEADER_NAME]: csrftoken,
+      [CSRF_HEADER_NAME]: csrf_token,
       "Content-type": "application/json",
     },
   };
 
   // The endpoint returns the updated project
-  const res = await fetch(endpoint, {
+  const resp = await fetch(endpoint, {
     ...options,
     body: JSON.stringify({ pinned }),
+  }).catch(console.error);
+
+  if (!resp) {
+    throw new Error("API error");
+  }
+
+  if (isErrorCode(resp.status)) {
+    throw new Error(resp.statusText);
+  }
+
+  return resp.json();
+}
+
+// writable methods
+/**
+ * Create a new project
+ *
+ * @param project
+ * @param csrf_token
+ * @param fetch
+ */
+export async function create(
+  project: {
+    title: string;
+    description?: string;
+    private?: boolean;
+    pinned?: boolean;
+  },
+  csrf_token: string,
+  fetch = globalThis.fetch,
+) {
+  const endpoint = new URL("projects/", BASE_API_URL);
+  const resp = await fetch(endpoint, {
+    body: JSON.stringify(project),
+    credentials: "include",
+    headers: {
+      "Content-type": "application/json",
+      [CSRF_HEADER_NAME]: csrf_token,
+      Referer: APP_URL,
+    },
+    method: "POST",
+  }).catch(console.error);
+
+  if (!resp) {
+    throw new Error("API unavailable");
+  }
+
+  if (isErrorCode(resp.status)) {
+    const { data } = await resp.json();
+    throw new Error(data);
+  }
+
+  return resp.json();
```

Comment:
It looks like this logic is being repeated in multiple different API handlers—might be ripe for a utility fn. Something like:

```typescript
/**
 * Returns data from a response or dies trying.
 *
 * @param resp The response to a `fetch` request.
 * @returns The data from a valid response body.
 * @throws Error if resp is void or contains a HTTP error code.
 */
async function getJsonData(resp?: Response) {
  if (!resp) {
    throw new Error("API unavailable");
  }

  if (isErrorCode(resp.status)) {
    const { data } = await resp.json();
    throw new Error(data);
  }

  return resp.json();
}
```

that can be invoked in API calls like:

```typescript
const resp = await fetch(...);
return getJsonData(resp);
```

I think this feedback can be addressed in a focused cleanup PR, since it may apply other places unchanged in this PR.

Reply:
More for #502